Added a very basic version of analytics for AMP.

avimehta opened the pull request:
This is the first PR in a series of PRs to implement the spec  outlined in #871.

In this commit, the tag parses the inline config and uses one builtin type to listen for
document.load events and sends a GET request when that happens. There is
primitive support for request templates and platform variables.

Things that are still to come:
- Support for more platform variables (future PR)
- Better instrumentation class (future PR)
- User specified variables (future PR)
- Readme.md (Future PR)

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.css:
```diff
@@ -0,0 +1,24 @@
+ * limitations under the License.
+ */
+
+amp-analytics {
```
This is too late. We have to move this definition to amp.css.

avimehta replied:
done in a previous commit.

dvoytenko commented on examples/analytics.amp.html:
```diff
@@ -0,0 +1,86 @@
+  {
+    "host": "google-analytics.com",
+    "requests": {
+      "default": "blah blah blah"
```
Please provide a meaningful value.

avimehta replied:
done.

dvoytenko commented:
Please reference in the description:
1. Issue#
2. Spec

dvoytenko commented on examples/analytics.amp.html:
```diff
@@ -0,0 +1,86 @@
+      {
+        "on": "TAP",
+        "request": "default",
+        "selector": ""
```
What does empty selector do?

avimehta replied:
undefined behavior for now. I removed it since the implementation for that part is not ready yet. Once we start working on it, we can define what empty selector does.

dvoytenko commented on examples/analytics.amp.html:
```diff
@@ -0,0 +1,86 @@
+    },
+    "triggers": [
+      {
+        "on": "TAP",
```
Do events have to be capitalized?

avimehta replied:
nope. Outdated example.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+ */
+
+import {Layout} from '../../../src/layout';
+// import * as st from '../../../src/style';
```
please remove

avimehta replied:
removed.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+
+export class AmpAnalytics extends AMP.BaseElement {
+
+  constructor(element) {
```
Please don't use constructor and firstAttachedCAllback - just put everything in the buildCallback - for all purposes it is the constructor.

avimehta replied:
done.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.FIXED;
```
Sounds like you can accept any layout out there. Just return `true`.

avimehta replied:
ack.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+    let config = this.element.textContent;
+    // TODO: Add try catch and other error conditions below.
+    config = JSON.parse(config);
+    this.config_ = config;
```
```
this.config_ = JSON.parse(this.element.textContent);
```

Once you move it to buildCallback, this will be sufficient. All errors are already properly handled, so that TODO can be removed as well.

avimehta replied:
done.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+
+  constructor(element) {
+    super(element);
+    this.config_ = undefined;
```
Please define types for each.

avimehta replied:
done.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+    this.config_ = config;
+
+    const type = this.element.getAttribute('type');
+    if (type && AnalyticsConfig[type]) {
```
Cross-module map access should be redone as a getter. E.g. `getAnalyticsConfig(type)` or `hasAnalyticsConfig(type)`.

avimehta replied:
the eventual goal was to make the vendor config JSON. @cramforce does this look good to you?

dvoytenko replied:
Internally they can be stored as JSON - that's fine. Externally is a different story.

dvoytenko commented on extensions/amp-analytics/0.1/vendors.js:
```diff
@@ -0,0 +1,38 @@
+ * limitations under the License.
+ */
+
+export const AnalyticsConfig = {
```
Type definition and typedef for the type.

dvoytenko replied:
Protect the map from cross-module access via getter.

avimehta replied:
Let me know if this is still needed.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+      this.type_ = type;
+
+      // Handle Optout.
+      if (AnalyticsConfig[type]['optout'] &&
```
Access to 'optout' should be declared via typedef and access without quotes here.

dvoytenko replied:
Alternatively, you can keep quotes but then you have to add quotes in the AnalyticsConfig properties. Basically it should be the same here and there.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+    this.generateRequests_();
+
+    if (!Array.isArray(config['triggers'])) {
+      // TODO: Nothing to do here. Show error in dev mode.
```
Implement TODO right away.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+    // Trigger callback can be synchronous. Do the registration at the end.
+    for (const trigger of config['triggers']) {
+      if (!trigger['on'] || !trigger['request']) {
+        // TODO: Show warning in dev mode.
```
ditto

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+  /** @override */
+  buildCallback() {
+    // Remove user defined size. Pixels should always be the default size.
+    this.element.style.width = '';
```
You already have `!important` so this is unneeded.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+      if (!config || !config['requests']) {
+        continue;
+      }
+      for (const name in config['requests']) {
```
Please use the same var name here and below, e.g. "k"

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+  /**
+   * Replaces template variables corresponding to a request with the value of
+   * that request. If no request is found, an empty string is inserted instead.
+   */
```
Define params and return type.

dvoytenko replied:
Also `@private`

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+    if (depth <= 0) {
+      return '';
+    }
+    const that = this;
```
Don't need this. Use `() => {}` functions that preserve `this`.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+    const that = this;
+    depth--;
+    const request = formatString.replace(expression, function(match, name) {
+      return that.expandRequest_(requests[name], expression, requests, depth) ||
```
Could you pls clarify why this is recursive?

avimehta replied:
For a setup like below:


'requests' : {
  'base': '/something',
  'pageview': '{base}?domain={domain}',
  'pageview_with_custom_dimensiom': '{pageview}&cd1={custom_dimension}'
}

The first two requests could be in predefined type with the third defined in inline config.

dvoytenko replied:
SG

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+  /**
+   * Replaces template variables corresponding to variables with it's value
+   * provided by the trigger, the config and the  platform in that order.
+   */
```
Type definitions for params and return

dvoytenko replied:
Also `@private`

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+
+  /**
+   * Callback for events that are registered by the config's triggers. This
+   * method generates the request and sends the request out.
```
ditto on types/private

avimehta replied:
done.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+  fillVariables_(formatString) {
+
+    // TODO: Support for other variables.
+    const vars = platformVariablesFor(this.element.ownerDocument.defaultView);
```
Use `this.getWin()` to get the window.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+   * method generates the request and sends the request out.
+   */
+  handleEvent_(trigger, event) {
+    const baseConfig = AnalyticsConfig[this.type_] || {};
```
Would be easier to merge this.config_ and base config in the initialization and then just use one config throughout?

avimehta replied:
Done this now. Included code from another repo. Let me know if there are any licensing issues. (i think there aren't). The original author ( @bnkuhn ) is also part of AMPProject.

dvoytenko replied:
Don't see any changes in this regard. Could you point us exactly where? What do you mean "included code from another repo"?

@erwinmombay to check on external dependencies.

avimehta replied:
https://github.com/avimehta/amphtml/commit/2eb3094cca62f9fd6ae4b905ead090f5bb91c36d#diff-566506fb6e7fab7f8f059ba89e02e02d 

I copied the code from the other repo and removed stuff that wasn't needed.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+
+    request = this.fillVariables_(request);
+
+    // TODO: Support for sendBeacon.
```
Is sendBeacon really applicable here? If so, why defer this work? If you still think it should be deferred - create an issue and reference it in this TODO. Format is:
`TODO(username, #ISSUE): ...`

avimehta replied:
Yes, using sendbeacon is very usefule on unloads/hidden. In general, some vendors might want to use it all the time.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+    }
+  }
+
+  sendRequest(request) {
```
Types

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+
+  sendRequest(request) {
+    // TODO: Logging should happen in dev mode only.
+    console./*OK*/info(this.getName() + ': Sent Request: ' + request);
```
Check devmode.

dvoytenko replied:
I.e. solve this TODO right away.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+    image.width = 1;
+    image.height = 1;
+    // Make it take zero space
+    this.element.style.width = 0;
```
No need. It's already configured right.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+  }
+
+  getName() {
+    return 'AmpAnalytics - ' + this.element.getAttribute('id');
```
What if ID is not available?

avimehta replied:
Don't know. Made some updates. Does it look better now?

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+    this.element.appendChild(image);
+  }
+
+  getName() {
```
Types

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+    const image = new Image();
+    image.src = request;
+    image.width = 1;
+    image.height = 1;
```
In devmode you need to handle onload/onerror (see event-helper.js for help) and log things accordingly.

dvoytenko commented on extensions/amp-analytics/0.1/instrumentation.js:
```diff
@@ -0,0 +1,71 @@
+import {getService} from '../../../src/service';
+import {documentStateFor} from '../../../src/document-state';
+import {Observable} from '../../../src/observable';
+
```
Types everywhere.

dvoytenko commented on extensions/amp-analytics/0.1/instrumentation.js:
```diff
@@ -0,0 +1,71 @@
+
+  addListener(eventType, listener) {
+
+    // TODO: Add support for clicks, timers, scroll etc.
```
TODO format + issue#

dvoytenko commented on extensions/amp-analytics/0.1/instrumentation.js:
```diff
@@ -0,0 +1,71 @@
+  return getService(window, 'amp-analytics-instrumentation', () => {
+    return new InstrumentationService(window);
+  });
+};
```
no semicolon here.

erwinmombay commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,190 @@
+    }
+
+    // Trigger callback can be synchronous. Do the registration at the end.
+    for (const trigger of config['triggers']) {
```
we currently try and avoid `for...of` since it generates 2-3 try catch blocks.

avimehta replied:
I tried searching but didn't find concrete details. Can you give me some context? (this sounds interesting). Changed for...of to for...in. Does this work?

dvoytenko commented on extensions/amp-analytics/0.1/instrumentation.js:
```diff
@@ -0,0 +1,71 @@
+  addListener(eventType, listener) {
+
+    // TODO: Add support for clicks, timers, scroll etc.
+    if (eventType.toLowerCase() == 'visible') {
```
Let's do const enum for event types.

dvoytenko replied:
Fine for now, but when we add new event handlers - we need to restructure to do something like `Map<eventType, registrationFactory>`

avimehta replied:
This whole class is going to change a lot. I don't expect any of this code to stay. I put it in place just to ensure that something works... @btownsend will be making most of the future changes here.

dvoytenko commented on extensions/amp-analytics/0.1/instrumentation.js:
```diff
@@ -0,0 +1,71 @@
+
+/**
+ * @param {!Window} window
+ * @return {!ClientIdService}
```
Wrong return type.

dvoytenko commented on extensions/amp-analytics/0.1/instrumentation.js:
```diff
@@ -0,0 +1,71 @@
+      if (this.docState_.getVisibilityState() == 'visible') {
+        listener(new AnalyticsEvent('visible'));
+      } else {
+        this.docState_.onVisibilityChanged(() => {
```
DocState is not a good fit here. We need to use `Viewer.isVisible` and `Viewer.onVisibilityChanged`.

avimehta replied:
Will update this when the class is re-written?

dvoytenko replied:
No. Let's do it right away. This is a very small change.

avimehta replied:
done.

dvoytenko commented on extensions/amp-analytics/0.1/instrumentation.js:
```diff
@@ -0,0 +1,71 @@
+import {Observable} from '../../../src/observable';
+
+export function addListener(eventType, listner) {
+  return instrumentationServiceFor(window).addListener(eventType, listner);
```
You need to pass `window` as an argument to `addListener`.

dvoytenko commented on extensions/amp-analytics/0.1/platform-variables.js:
```diff
@@ -0,0 +1,100 @@
+
+import {getService} from '../../../src/service';
+import {documentInfoFor} from '../../../src/document-info';
+import {parseUrl, removeFragment} from '../../../src/url';
```
Types, docs

avimehta replied:
done.

dvoytenko commented on extensions/amp-analytics/0.1/platform-variables.js:
```diff
@@ -0,0 +1,100 @@
+    /**
+     * Returns a random value for cache busters.
+     */
+    'random': () => {
```
Why not just use UrlReplacements class? I think it covers all of these. No?

avimehta replied:
Right now they are pretty much the same but I intend to add a lot more stuff here that is not in URL replacement class. If all the variables defined in the spec are okay for url-replacement class, let me know and I can switch over to that. Another  issue is with the templating format. We would prefer the templates to be a little more than just uppercase words. I added {} for that reason. I can change it to {{}} or something else. If url-replacement can be configured to support {} or {{}} then let me know and I'll remove this.

dvoytenko replied:
UrlRepalcements is extensible - you can add any additional mappings you want. Our preference was for upper case so far. We can support both options. But I don't think we should have competing formats for URL substitutions in our library. I.e. we would ideally make a decision on one and stick with it.

avimehta replied:
I switched to using url-replacement for now. I think we'll have to make some changes to it before it is usable for amp-analytics.

btownsend replied:
Although url-replacements is extensible, a design where tokens are simply uppercase strings will run into problems if one token identifier contains a substring of any other token identifier. Since we allow publishers to specify their own tokens (in the amp-analytics config / trigger config) we're going to run into that case eventually.

I haven't thought this through very deeply, but on the surface it seems easier to implement a system where tokens can be identified in a template string without knowledge of all possible token identifiers. E.g., if tokens are contained within curly braces, you can do something like this:

  templateString.replace(/{([^}]*)}/g, (m) => { return dataModel.get(m) });

(vs. the concatenation of all identifiers approach in url-replacement.)

dvoytenko commented on extensions/amp-analytics/0.1/test/test-amp-analytics.js:
```diff
@@ -0,0 +1,56 @@
+    const el = document.createElement('amp-analytics');
+    el.textContent = config;
+    const analytics = new AmpAnalytics(el);
+    analytics.firstAttachedCallback();
```
Since AmpAnalytics listens to the real events, shouldn't you also remove it from DOM or deconstruct somehow to avoid listener leaks and tinted tests?

avimehta replied:
I did not address any of the feedback from this class. I am still working on it so expect comments to be addressed in my next commit.

avimehta replied:
Done now. is anything else needed?

dvoytenko commented on extensions/amp-analytics/0.1/test/test-amp-analytics.js:
```diff
@@ -0,0 +1,56 @@
+      'triggers': [trigger]
+    });
+
+    expect(analytics.element.getElementByTagName('img')).to.have.length(0);
```
Why do you expect "img" elements inside? I didn't see those ever being added. Am I wrong?

dvoytenko replied:
There's no such method getElementByTagName

avimehta replied:
I do add the img tags to the amp-analytics tag. 

git should have been getElementsByTagName.

dvoytenko commented on extensions/amp-analytics/0.1/test/test-amp-analytics.js:
```diff
@@ -0,0 +1,56 @@
+    expect(els[0].getAttribute('src')).to.equal('https://example.com/bar');
+  });
+
+  // TODO: Add tests for:
```
Please add these tests.

avimehta replied:
added most of them. Two of them are disabled right now because I do not know how to change the value of AnalyticsConfig to something that I can test with. Ideas?

dvoytenko commented on extensions/amp-analytics/0.1/vendors.js:
```diff
@@ -0,0 +1,38 @@
+ */
+
+export const AnalyticsConfig = {
+  // TODO: Add a generic hit format to make custom analytics easier.
```
TODO format

avimehta replied:
done.

avimehta commented:
ptal. I have not updated the tests file because that will take some more time. Meanwhile, I wanted to get feedback on rest of the stuff.

dvoytenko commented on extensions/amp-analytics/0.1/vendors.js:
```diff
@@ -0,0 +1,43 @@
+ */
+
+/**
+ * @const {JSONObject}
```
!JSONObject

avimehta replied:
done.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,253 @@
+    // TODO(btownsend, #871): Add support for variables from inline config.
+    const vars = platformVariablesFor(this.getWin());
+    const expr = platformVariablesExpressionFor(
+        this.element.ownerDocument.defaultView);
```
`this.getWin()`

avimehta replied:
Removed this code.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,253 @@
+   * @private
+   */
+  getName_() {
+    return 'AmpAnalytics ' + this.element.getAttribute('id') || '<unknown id>';
```
You need to put `||` expression in the parens.

avimehta replied:
Done.

Huanzhang89 commented:
When will this get merged into master? Or is there a way I can use this version of the amp source code to test running google analytics?

Thanks!

dvoytenko commented:
@avimehta not sure if this is ready to be reviewed again. I see many of the previous comments not responded to. LMK.

avimehta commented:
@dvoytenko I think I have now addressed all the comments. Please let me know if I missed something. 

The extension is currently not experimental. Is there a way I can mark it as experimental so that people don't start using it? 

Overall, I would like to merge this asap if the major stuff is good. We can fix the smaller issues in a future PR if needed.

powdercloud commented on examples/analytics.amp.html:
```diff
@@ -0,0 +1,79 @@
+  {
+    "host": "my-analytics.com",
+    "requests": {
+      "default": "/log?domain=DOMAIN&path=PATH"
```
In the source view, github highlights the '&' character in this line. Should this be &amp; ?

dvoytenko replied:
Yes - it's an issue. It has to be `&amp;`. This is the reason I wanted to do this via `<script>` element. Otherwise we have to do escapes. @cramforce fyi.

cramforce replied:
Escaping sounds good to me.

btownsend replied:
I'm concerned escaping HTML in something that looks like JSON will not be intuitive for publishers. What would the script element approach look like? Are there any downsides?

dvoytenko replied:
It will look like this:

```html
<amp-analytics id="analytics1">
  <script type="application/ld+json">
   {
    "host": "my-analytics.com",
    ...
  }
  </script>
</amp-analytics>
```

dvoytenko replied:
I don't see a problem with this to be honest. We could always support both forms as well, but the cost of one `script` tag seems very low to me. Not sure what this entail for the validator, but we already use `script` for this purpose for JSON metadata.

cramforce replied:
It isn't json-ld, but we can whitelist a new non-script type. I'd vote for
just "json".
On Dec 4, 2015 2:04 PM, "Dima Voytenko" <notifications@github.com> wrote:

> In examples/analytics.amp.html
> <https://github.com/ampproject/amphtml/pull/982#discussion_r46718647>:
>
> > +      max-height: 80%;
> > +      width: 100%;
> > +    }
> > +  </style>
> > +  <script async custom-element="amp-analytics" src="https://cdn.ampproject.org/v0/amp-analytics-0.1.js"></script>
> > +  <style>body {opacity: 0}</style><noscript><style>body {opacity: 1}</style></noscript>
> > +  <script async src="https://cdn.ampproject.org/v0.js"></script>
> > +</head>
> > +<body>
> > +
> > +<article>
> > +  <amp-analytics id="analytics1">
> > +  {
> > +    "host": "my-analytics.com",
> > +    "requests": {
> > +      "default": "/log?domain=DOMAIN&path=PATH"
>
> I don't see a problem with this to be honest. We could always support both
> forms as well, but the cost of one script tag seems very low to me. Not
> sure what this entail for the validator, but we already use script for
> this purpose for JSON metadata.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ampproject/amphtml/pull/982/files#r46718647>.
>

dvoytenko replied:
Sorry, yes, I mean just "application/json". This is the standard mime type for JSON. See http://www.ietf.org/rfc/rfc4627.txt.

avimehta replied:
I would like to get this PR merged soonish. @btownsend will take care of whatever you guys decide in his next PR.

dvoytenko replied:
Please create an issue and assign appropriately then.

avimehta replied:
#1081

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,232 @@
+import {addListener} from './instrumentation';
+import {AnalyticsConfig} from './vendors';
+
+export class AmpAnalytics extends AMP.BaseElement {
```
See here on how to implement experiment control: https://github.com/ampproject/amphtml/blob/master/extensions/amp-mustache/0.1/amp-mustache.js#L51

You will also need to add the experiment definition to here:
https://github.com/ampproject/amphtml/blob/master/tools/experiments/experiments.js#L42

avimehta replied:
done.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,232 @@
+    // TODO(btownsend, #871): Add support for variables from inline config.
+    request = urlReplacementsFor(this.getWin()).expand(request);
+
+    // TODO(btownsend, #871): Add support for sendBeacon.
```
Please create a separate issue for this. E.g. "Use sendBeacon for amp-analytics".

avimehta replied:
done. #1061

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,232 @@
+   * @private
+   */
+  mergeConfigs_() {
+    // TODO(btownsend, #871): Implement support for remote configuration.
```
A separate issue? It's hard to just pack every feature request into one issue.

avimehta replied:
This is pretty important to spec. will keep it as it is.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,232 @@
+import {Layout} from '../../../src/layout';
+import {log} from '../../../src/log';
+import {loadPromise} from '../../../src/event-helper';
+import {mergeObjects} from '../../../third_party/data-layer-helper/helper';
```
@erwinmombay wdyt? Given what it does, I don't see a reason to add a 3rd party dependency here - too easy to do it ourselves. No?

erwinmombay replied:
agree. theres some extra code there for "non Plain" object which i doubt we need.

dvoytenko replied:
Yes - we only have JSON objects, so this should be simpler.

avimehta replied:
done.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,232 @@
+    /**
+     * @private {?JSONValue} The analytics config associated with the tag
+     */
+    this.config_ = null;
```
It's used everywhere as a non-nullable value. Can you redefine this code to make it clear and change typing? Also, I think it should be `!JSONObject`

dvoytenko replied:
For instance, you could do:

```
this.config_ = this.mergeConfigs_();
```

avimehta replied:
done.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,232 @@
+    this.type_ = null;
+
+    /**
+     * @private {Object.<string, string>} A map of request names to the request
```
!Object and no dots before `<`

avimehta replied:
done.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,232 @@
+    // TODO(btownsend, #871): Implement support for remote configuration.
+    const remote = {};
+
+    const inline = JSON.parse(this.element.textContent) || {};
```
Anything you need to do if parse fails?

avimehta replied:
added a warning. if some necessary stuff is missing, the tag will fail later.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,232 @@
+      obj = obj[props[i]];
+    }
+    return obj();
+
```
nit: extra line

avimehta replied:
done.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,232 @@
+      }
+      obj = obj[props[i]];
+    }
+    return obj();
```
reads strange: you are calling an object. Especially since you initialize it above as a window. I know it's correct, but perhaps some naming could be fixed?

avimehta replied:
didn't know what to rename to.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,232 @@
+    return obj();
+
+  }
+  /**
```
nit: add a line

avimehta replied:
done.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,232 @@
+   * that request. If no request is found, an empty string is inserted instead.
+   *
+   * @param {string} formatString The request string to be expanded
+   * @param {RegExp} expression The regular expression used to detect request
```
!RegExp

avimehta replied:
done.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,232 @@
+   * @param {string} formatString The request string to be expanded
+   * @param {RegExp} expression The regular expression used to detect request
+   *          templates
+   * @param {Object.<string, string> requests Map of requests to lookup template
```
No dots in types and '!'

avimehta replied:
done.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,232 @@
+   */
+  handleEvent_(trigger, event) {
+    const baseConfig = AnalyticsConfig[this.type_] || {};
+    const host = this.config_['host'] || baseConfig['host'];
```
So, why are we still doing this if it was merged above?

avimehta replied:
missed removing it. removed

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,232 @@
+    image.src = request;
+    image.width = 1;
+    image.height = 1;
+    this.element.appendChild(image);
```
Is appending image necessary?

avimehta replied:
nope. changed tests and removed this.

dvoytenko commented on extensions/amp-analytics/0.1/amp-analytics.js:
```diff
@@ -0,0 +1,232 @@
+    return 'AmpAnalytics ' +
+        (this.element.getAttribute('id') || '<unknown id>');
+  }
+};
```
nit: no need for semicolons.

avimehta replied:
ack.

dvoytenko commented on extensions/amp-analytics/0.1/instrumentation.js:
```diff
@@ -0,0 +1,102 @@
+import {Observable} from '../../../src/observable';
+
+/**
+ * @callback AnalyticsEventListener
```
Could you please do this via typedef instead? We don't use callback notation.

avimehta replied:
done.

dvoytenko commented on extensions/amp-analytics/0.1/instrumentation.js:
```diff
@@ -0,0 +1,102 @@
+
+/**
+ * @callback AnalyticsEventListener
+ * @param {AnalyticsEvent} event that got triggered.
```
'!'

avimehta replied:
changed

dvoytenko commented on extensions/amp-analytics/0.1/instrumentation.js:
```diff
@@ -0,0 +1,102 @@
+ */
+
+/**
+ * @param {Window} window Window object to listen on.
```
'!'

avimehta replied:
done

dvoytenko commented on extensions/amp-analytics/0.1/instrumentation.js:
```diff
@@ -0,0 +1,102 @@
+
+/**
+ * @param {Window} window Window object to listen on.
+ * @param {!Event} type Event type to listen to.
```
it's a string, right?

avimehta replied:
yes. didn't know how jsdoc worked for enums.

dvoytenko replied:
I was reacting to the global Event type in JS. With the new name, you can certainly use enum type here - that would be better.

dvoytenko commented on extensions/amp-analytics/0.1/instrumentation.js:
```diff
@@ -0,0 +1,102 @@
+ * @const
+ * @enum {string}
+ */
+export const Event = {
```
Let's rename - there's a global Event type already.

avimehta replied:
done.

dvoytenko commented on extensions/amp-analytics/0.1/instrumentation.js:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  /**
+   * @param {!Layout} eventType The type of event
```
string?

avimehta replied:
copy-paste error

dvoytenko commented on extensions/amp-analytics/0.1/instrumentation.js:
```diff
@@ -0,0 +1,102 @@
+   * @param {!Window} window
+   */
+  constructor(window) {
+    this.viewer_ = viewerFor(window);
```
`@const {!Viewer}`

avimehta replied:
done.

dvoytenko commented on extensions/amp-analytics/0.1/vendors.js:
```diff
@@ -0,0 +1,43 @@
+/**
+ * @const {!JSONObject}
+ */
+export const AnalyticsConfig = {
```
Ok. In this case, let's rename it as a constant. E.g. "ANALYTICS_CONFIG"

avimehta replied:
done.

dvoytenko commented on extensions/amp-analytics/amp-analytics.md:
```diff
@@ -0,0 +1,97 @@
+{
+  "host": "my-analytics.com:8080",
+  "requests": {
+    "base_hit": "/collect?v=1&_v=a0&aip=true&_s=HIT_COUNT&dl=DOMAIN&dt=TITLE&sr=SCREEN_WIDTHxSCREEN_HEIGHT&ht=TIMESTAMP&cid=CLIENT_IDENTIFIER&tid=ACCOUNT"
```
Link https://github.com/ampproject/amphtml/blob/master/spec/amp-var-substitutions.md somewhere in this doc.

avimehta replied:
added.

dvoytenko commented on third_party/data-layer-helper/LICENSE:
```diff
@@ -0,0 +1,14 @@
+Copyright 2012 Google Inc. All Rights Reserved.
```
Let's see if we can do without this. If we do still need it - we will need to submit it in a separate PR.

avimehta replied:
removed.

dvoytenko commented:
Cool. I think we are pretty close. But one request: could you please respond to each comment we make? It makes it a lot easier to review.

dvoytenko commented:
Good. So, two more actions:

1. Squash commits
2. Please make sure you create the issue for the `<script>` thing.

Otherwise, LGTM.

dvoytenko commented:
LGTM

rudygalfi commented:
:+1:

cramforce commented:
Woot.
